src/ws/cockpitchannelresponse.c

```diff
@@ -41,22 +44,37 @@ cockpit_channel_inject_free (gpointer data)
       if (inject->service)
         g_object_remove_weak_pointer (G_OBJECT (inject->service), (gpointer *)&inject->service);
       g_free (inject->base_path);
+      g_free (inject->host);
       g_free (inject);
     }
 }
 
 static CockpitChannelInject *
 cockpit_channel_inject_new (CockpitWebService *service,
-                            const gchar *path)
+                            const gchar *path,
+                            const gchar *host)
 {
   CockpitChannelInject *inject = g_new (CockpitChannelInject, 1);
   inject->service = service;
   g_object_add_weak_pointer (G_OBJECT (inject->service), (gpointer *)&inject->service);
   inject->base_path = g_strdup (path);
+  inject->host = g_strdup (host);
   return inject;
 }
 
 static void
+cockpit_channel_inject_update_checksum (CockpitChannelInject *inject,
+                                        GHashTable *headers)
+{
+  const gchar *checksum = g_hash_table_lookup (headers, CHECKSUM_HEADER);
+
+  if (checksum)
+    cockpit_web_service_set_host_checksum (inject->service, inject->host, checksum);
+
+  g_hash_table_remove (headers, CHECKSUM_HEADER);
```

Comment:
I dont' understand why this should modify `headers`? Is that only to avoid updating it again, in case it gets called twice (which would still be a no-op, as set_host_checksum already compares the checksum)? This still looks potentially dangerous, as you then lose the information in the header. This should  at least get a comment.